### ISSUE TITLE
Automatic EULA generation failing for Apache-2.0

### ISSUE THREAD

**Issue by zacps** [volks73/cargo-wix#71]
Hi,
I've started trying to add support for this to Alacritty but I can't get the automatic EULA generation to work.

Alacritty is licensed under `Apache-2.0` which it seems you have support for, however when running the installer the EULA field is blank:
![image](https://user-images.githubusercontent.com/9827696/47549690-553a4900-d959-11e8-9032-ba3bd6e190e1.png)

My current progress is [here](https://github.com/zacps/alacritty/tree/cargo-wix) and I'm running with:

```
cargo wix assets\windows\msi.wxs --nocapture -l LICENSE-APACHE
```

**Comment by volks73** [comment]
I believe the problem is with the `-l LICENSE-APACHE` option. Try removing it from the invocation. For example:

```dos
C:\> cargo wix assets\windows\msi.wxs --nocapture
```

The problem is that the End User License Agreement (EULA) file for the license dialog must be in the [Rich Text Format (RTF)](https://en.wikipedia.org/wiki/Rich_Text_Format). I believe your `LICENSE-APACHE` is just a plain text file with no markup or formatting. So, the WiX Toolset compiler (candle) and linker (light) are working correctly because the file exists, but the RTF renderer within the installer fails to render the plain text file and yields an empty text.

If you do not specify the license with the `-l,--license` option, then the `license` field in the Cargo.toml file is used. Assuming the the `license` field is set to `Apache-2.0`, then cargo-wix will automatically generate a `License.rtf` file in the same location as the WXS file and be "linked" with the installer. The Apache-2.0 content should then appear in the installer. This is assuming you started with the `cargo wix --init` command without any further options or flags.

The `-l,--license` option is intended as a mechanism to use a custom license or a license not currently supported for automatic generation by cargo-wix and temporarily override a license if a one-off installer with a different license/EULA is needed. The assumption is that if you are using the `-l,--license` file, then you are on your own and may need to modify the WXS file manually, which can be done with any text editor as it is just an XML file. 

The use of the `-l,--license` option and confusion related to the EULA is recognized, and it currently being addressed. See the `master` branch of the repository, #49, and #54.

**Comment by zacps** [comment]
```
C:\Users\Zac\Programming\alacritty>cargo wix assets\windows\msi.wxs --nocapture --no-build
 WARN: Skipped building the release binary
Windows Installer XML Toolset Compiler version 3.11.1.2318
Copyright (c) .NET Foundation and contributors. All rights reserved.

msi.wxs
Windows Installer XML Toolset Linker version 3.11.1.2318
Copyright (c) .NET Foundation and contributors. All rights reserved.

C:\Users\Zac\Programming\alacritty\assets\windows\msi.wxs(56) : error LGHT0103 : The system cannot find the file 'LICENSE'.
Error[1] (Command): The 'light' application failed with exit code = 103. Consider using the '--nocapture' flag to obtain more information.
```

Doing that fails because the license file is named `LICENSE-APACHE`, not `LICENSE`. There's also a minor bug where is recommends running with `--nocapture` when I am already doing so.

**Comment by volks73** [comment]
I understand you may just want this to work and build an installer without knowing all of the gory details of the implementation and reasoning, so please feel free to skip to one of the following three sections that include possible resolutions for your issue:

1. [Manually Edit](#manually-edit)
2. [Start over](#start-over)
3. [Upgrade](#upgrade)

It would be great if you could report back which option you chose and any modifications or tweaks that might be needed in case someone else encounters a similar problem.

## Background

Okay, there are a couple of things at work here: 

First, the license dialog is blank in the installer because in the `msi.wxs` file for your package at line 125, the `Value` attribute for the `WixVariable` node is `LICENSE-APACHE`. The `LICENSE-APACHE` file is not a RTF file, so the contents of the license dialog will be blank. A RTF file needs to be created. 

Second, when the `-l,--license` option is dropped from the invocation of the cargo-wix subcommand, the assumption is that the license file, which should be used for the [sidecar](https://en.wikipedia.org/wiki/Sidecar_file) license file, is a `LICENSE` file found in the root of the package's folder tree, i.e. the same folder as the package's manifest (Cargo.toml). The initial implementation of the cargo-wix subcommand included a survey of many of the popular packages on [crates.io](https://crates.io/) and their respective organization. The majority of surveyed projects had a single LICENSE file in the root folder, even if the package was multi-licensed, such as MIT/APACHE-2.0. The convention appeared to be to have three files for a multi-licensed package: 

#. LICENSE
#. LICENSE-MIT
#. LICENSE-APACHE 

Thus, the initial implementation of cargo-wix looked for a LICENSE file in the root folder if the `-l,--license` option was not used since this would include all of the licenses. This default behavior has lead to the error message you are receiving. 

The `-l,--license` option only pertained to the sidecar license file, not the EULA in the license dialog of the installer. The implementation in v0.0.5 does not include any way to define a EULA other than the ID from the `license` field in the package's manifest (Cargo.toml). Again, the survey suggested that most projects used one of three licenses: MIT, Apache-2.0, or GPL-3.0 and did _not_ include a RTF license file that could be used for the EULA in the installer. The assumption was if a project was _not_ using one of these three licenses, the developer would be aware and responsible for creating a RTF license on his or her own and would have to manually alter the EULA in the WXS file. The majority use-case would be the defaults: generate a RTF license file based on the license ID in the Cargo.toml file and place it along side the WXS file during initialization. If during creation/building of the installer a separate sidecar license was desired, the `-l,--license` could be used.

Again, the confusion and weirdness with the License files, EULAs, and CLI options has been identified and it is currently being addressed. I apologize for the lack of polish and smoothness in the alpha version of this subcommand in this regard.

## Manually Edit

The WXS file is an XML file and can be modified with any reasonable text editor. However, first you need to generate a RTF license file to address the blank license dialog. As you have rightfully noticed, there is an embedded RTF template for the Apache-2.0 license in the cargo-wix subcommand. Complete the following steps to generate a `Eula.rtf` file for the installer:

1. Start the command prompt (cmd.exe).
2. Navigate to the root folder of your project.
3. Execute the following command to generate a `Eula.rtf` file for the Apache-2.0 license:

```dos
C:\Path\To\Project> cargo wix --print-template Apache-2.0 > assets\windows\Eula.rtf 
```

4. Open the `msi.wxs` file in a suitable text editor.
5. Navigate to Line 125.
6. Change the value of the `Value` attribute for the `WixVariable` node from `LICENSE-APACHE` to `assets\windows\Eula.rtf`, i.e. the line should look like this after modification:

```xml
<WixVariable Id='WixUILicenseRtf' Value='asserts\windows\Eula.rtf'/>
```

7. Navigate to Line 60.
8. Change the value for the `Source` attribute from `$(var.LicenseSource)` to `LICENSE-APACHE`, i.e. the line should look like this after modification:

```xml
Source='LICENSE-APACHE`
```

9. Save the `msi.wxs` file.

Once these steps are completed, you should be able to run the subcommand without any options to build/create an installer as follows:

```dos
C:\Path\to\Project> cargo wix assets\windows\msi.wxs
```

## Start Over

Since your project is licensed with the Apache-2.0 license, which is supported by the cargo-wix subcommand, the default behavior for the subcommand should "just work". Note, this option will erase any customizations you may have made to the `msi.wxs` file. I noticed an addition binary component added to the file. Complete the following steps to use the default behavior:

1. Start a command prompt.
2. Navigate to the root folder of your project.
3. Open the package's manifest (Cargo.toml) in a suitable text editor.
4. Ensure the `license` field exists for the `[package]` section and its value is `Apache-2.0`.
5. Save and close the package's manifest file.
6. Initialize a new WXS file with the following command:

```dos
C:\Path\to\Project> cargo wix --init
```

7. Ensure a `wix` folder has been created within the root folder of your project and that a `main.wxs` file exists.
8. Ensure a `License.rtf` exists within the created `wix` folder as well.
9. Build the installer with the following command:

```dos
C:\Path\to\Project> cargo wix
```

10. Test the installer and ensure the Apache-2.0 license content appears in the license dialog for the installer.
11. Complete or cancel the installer.
12. Modify the generated the `wix\main.wxs` with any customizations. If you move the location or rename the `main.wxs` file, then some modification to paths within the WXS file maybe needed and the `cargo wix` subcommand will have to be executed with the path to the WXS file as the first positional argument.

## Upgrade

I am currently hard at work with a major update on the `master` branch for this subcommand that (hopefully) removes many of these problems with the licenses and EULAs. The functionality has been implemented, I am currently testing, debugging, and documenting the changes and update. So, the `master` branch appears to be very usable and should resolve this issue you are having. You can install the git repository `master` version instead of the current released version as follows:

```dos
C:\> cargo install --force --git https://github.com/volks73/cargo-wix.git
```

This will overwrite the v0.0.5 release. Note, you may still have to perform Option 1 and/or 2 even with the upgrade and many of the options and flags have been turned into subcommands, but overall experience with the cargo-wix subcommand should be more polished. There is the `-e,--eula` and `-l,--license` options for the `cargo wix init` subcommand that you can use to re-initialize the WXS file without having to manually edit the output afterwards. You can also directly save the output of the `cargo wix init` subcommand to the `assets\windows` folder with the `-o,--output` option. However, the Wix section for the Cargo.toml file has not been implemented in this revision, so you will still need build the installer with the `cargo wix Path\to\File.wxs`. The focus is on adding polish and tests to ensure current features work as expected without weirdness.

**Comment by volks73** [comment]
I think origin of the original issue has been resolved with the recent release (v0.1.0) and its related documentation. The workarounds are still valid, but should not be needed.
